### DISCUSSION TITLE
4.0.2 crash in tr_peerMgrPeerStats()

### DISCUSSION THREAD

**Discussion by dantearmok** [transmission/transmission#5253] [Support]
Mac mini (late 2012, 2.5-GHz 4-Core i5), 16GB RAM, plus an assortment of USB & FW HDs.
macOS 10.15 Catalina, ClamXAV, PeakHour, and PIA (all fully updated).
Running Transmission (f5a24862db) -- March 17th's midnight nightly.

Transmission crashed in the background.

Process:               Transmission [5809]
Path:                  /Applications/Transmission.app/Contents/MacOS/Transmission
Identifier:            org.m0k.transmission
Version:               4.0.2-dev (14714.4.0.2)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Transmission [5809]
User ID:               502

Date/Time:             2023-03-18 14:04:00.167 -0400
OS Version:            Mac OS X 10.15.7 (19H2026)
Report Version:        12
Anonymous UUID:        9D83084F-22F2-20FB-B0D2-056E94905F2F

Time Awake Since Boot: 46000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000020
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [5809]

VM Regions Near 0x20:
--> 
    __TEXT                 000000010adba000-000000010aff2000 [ 2272K] r-x/r-x SM=COW  /Applications/Transmission.app/Contents/MacOS/Transmission

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   org.m0k.transmission          	0x000000010aec56a6 tr_peerMgrPeerStats(tr_torrent const*, unsigned long*) + 372
1   org.m0k.transmission          	0x000000010adfcb2e -[Torrent peers] + 59
2   org.m0k.transmission          	0x000000010ae4fa84 -[InfoPeersViewController updateInfo] + 1087
3   org.m0k.transmission          	0x000000010ae16200 -[InfoWindowController updateInfoStats] + 48
4   org.m0k.transmission          	0x000000010adcc35f -[Controller updateUI] + 922
5   org.m0k.transmission          	0x000000010adcc42f -[Controller fullUpdateUI] + 30
6   org.m0k.transmission          	0x000000010adce137 -[Controller torrentFinishedDownloading:] + 1597
7   com.apple.CoreFoundation      	0x00007fff35d69549 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
8   com.apple.CoreFoundation      	0x00007fff35d694dd ___CFXRegistrationPost1_block_invoke + 63
9   com.apple.CoreFoundation      	0x00007fff35d69452 _CFXRegistrationPost1 + 372
10  com.apple.CoreFoundation      	0x00007fff35d690be ___CFXNotificationPost_block_invoke + 80
11  com.apple.CoreFoundation      	0x00007fff35d3926c -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1745
12  com.apple.CoreFoundation      	0x00007fff35d38635 _CFXNotificationPost + 1586
13  com.apple.Foundation          	0x00007fff383de096 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
14  org.m0k.transmission          	0x000000010ae01bbb -[Torrent completenessChange:wasRunning:] + 309
15  org.m0k.transmission          	0x000000010adbd378 invocation function for block in onTorrentCompletenessChanged(tr_torrent*, tr_completeness, bool, void*) + 69
16  libdispatch.dylib             	0x00007fff6fa9a6c4 _dispatch_call_block_and_release + 12
17  libdispatch.dylib             	0x00007fff6fa9b658 _dispatch_client_callout + 8
18  libdispatch.dylib             	0x00007fff6faa6cab _dispatch_main_queue_callback_4CF + 936
19  com.apple.CoreFoundation      	0x00007fff35db2df3 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 9
20  com.apple.CoreFoundation      	0x00007fff35d72a8a __CFRunLoopRun + 2042
21  com.apple.CoreFoundation      	0x00007fff35d71c33 CFRunLoopRunSpecific + 466
22  com.apple.HIToolbox           	0x00007fff3498daad RunCurrentEventLoopInMode + 292
23  com.apple.HIToolbox           	0x00007fff3498d7c5 ReceiveNextEventCommon + 584
24  com.apple.HIToolbox           	0x00007fff3498d569 _BlockUntilNextEventMatchingListInModeWithFilter + 64
25  com.apple.AppKit              	0x00007fff32fd63c9 _DPSNextEvent + 883
26  com.apple.AppKit              	0x00007fff32fd4c10 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1352
27  com.apple.AppKit              	0x00007fff32fc691e -[NSApplication run] + 658
28  com.apple.AppKit              	0x00007fff32f98706 NSApplicationMain + 777
29  libdyld.dylib                 	0x00007fff6faf4cc9 start + 1

Thread 1:
0   org.m0k.transmission          	0x000000010aec6ac1 tr_peerMgr::reconnectPulse() + 1121
1   org.m0k.transmission          	0x000000010aec876b tr_peerMgr::bandwidthPulse() + 217
2   org.m0k.transmission          	0x000000010af56c4b event_process_active_single_queue + 1121
3   org.m0k.transmission          	0x000000010af53aeb event_base_loop + 1134
4   org.m0k.transmission          	0x000000010aeb4245 tr_session_thread_impl::sessionThreadFunc(event_base*) + 87
5   org.m0k.transmission          	0x000000010aeb4637 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*> >(void*) + 63
6   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
7   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 2:
0   libsystem_kernel.dylib        	0x00007fff6fc38746 __semwait_signal + 10
1   libsystem_c.dylib             	0x00007fff6fbbbeea nanosleep + 196
2   libc++.1.dylib                	0x00007fff6ce046da std::__1::this_thread::sleep_for(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > const&) + 73
3   org.m0k.transmission          	0x000000010aef14f0 tr_web::Impl::curlThreadFunc() + 448
4   org.m0k.transmission          	0x000000010aef1764 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (tr_web::Impl::*)(), tr_web::Impl*> >(void*) + 59
5   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
6   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 3:
0   libsystem_kernel.dylib        	0x00007fff6fc3a756 kevent + 10
1   org.m0k.transmission          	0x000000010ae63172 -[VDKQueue watcherThread:] + 203
2   com.apple.Foundation          	0x00007fff3842d222 __NSThread__start__ + 1064
3   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
4   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 4:: com.apple.NSEventThread
0   libsystem_kernel.dylib        	0x00007fff6fc35dea mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff6fc36160 mach_msg + 60
2   com.apple.CoreFoundation      	0x00007fff35d73d07 __CFRunLoopServiceMachPort + 247
3   com.apple.CoreFoundation      	0x00007fff35d727c1 __CFRunLoopRun + 1329
4   com.apple.CoreFoundation      	0x00007fff35d71c33 CFRunLoopRunSpecific + 466
5   com.apple.AppKit              	0x00007fff33177bf4 _NSEventThread + 132
6   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
7   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 5:: com.apple.coreaudio.AQClient
0   libsystem_kernel.dylib        	0x00007fff6fc35dea mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff6fc36160 mach_msg + 60
2   com.apple.CoreFoundation      	0x00007fff35d73d07 __CFRunLoopServiceMachPort + 247
3   com.apple.CoreFoundation      	0x00007fff35d727c1 __CFRunLoopRun + 1329
4   com.apple.CoreFoundation      	0x00007fff35d71c33 CFRunLoopRunSpecific + 466
5   com.apple.audio.toolbox.AudioToolbox	0x00007fff340e8f60 GenericRunLoopThread::Entry(void*) + 166
6   libAudioToolboxUtility.dylib  	0x00007fff6c47c611 CAPThread::Entry(CAPThread*) + 77
7   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
8   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 6:
0   libsystem_kernel.dylib        	0x00007fff6fc35e26 semaphore_wait_trap + 10
1   com.apple.audio.caulk         	0x00007fff69638b16 caulk::mach::semaphore::wait() + 16
2   com.apple.audio.caulk         	0x00007fff696389b2 caulk::semaphore::timed_wait(double) + 106
3   com.apple.audio.caulk         	0x00007fff696387c4 caulk::concurrent::details::worker_thread::run() + 30
4   com.apple.audio.caulk         	0x00007fff696381e4 void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*> > >(void*) + 45
5   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
6   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 7:: AMCP Logging Spool
0   libsystem_kernel.dylib        	0x00007fff6fc35e26 semaphore_wait_trap + 10
1   com.apple.audio.caulk         	0x00007fff69638b16 caulk::mach::semaphore::wait() + 16
2   com.apple.audio.caulk         	0x00007fff696389b2 caulk::semaphore::timed_wait(double) + 106
3   com.apple.audio.caulk         	0x00007fff696387c4 caulk::concurrent::details::worker_thread::run() + 30
4   com.apple.audio.caulk         	0x00007fff696381e4 void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*> > >(void*) + 45
5   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
6   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 8:
0   libsystem_pthread.dylib       	0x00007fff6fcf8b68 start_wqthread + 0

Thread 9:
0   libsystem_pthread.dylib       	0x00007fff6fcf8b68 start_wqthread + 0

Thread 10:
0   libsystem_pthread.dylib       	0x00007fff6fcf8b68 start_wqthread + 0

Thread 11:
0   libsystem_pthread.dylib       	0x00007fff6fcf8b68 start_wqthread + 0

Thread 12:: AQConverterThread
0   libsystem_kernel.dylib        	0x00007fff6fc38872 __psynch_cvwait + 10
1   libsystem_pthread.dylib       	0x00007fff6fcfd425 _pthread_cond_wait + 698
2   libAudioToolboxUtility.dylib  	0x00007fff6c49870c CAGuard::Wait() + 56
3   com.apple.audio.toolbox.AudioToolbox	0x00007fff3410d615 AQConverterManager::AQConverterThread::ConverterThreadEntry(void*) + 819
4   libAudioToolboxUtility.dylib  	0x00007fff6c47c611 CAPThread::Entry(CAPThread*) + 77
5   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
6   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 13:: com.apple.audio.IOThread.client
0   libsystem_kernel.dylib        	0x00007fff6fc35dea mach_msg_trap + 10
1   libsystem_kernel.dylib        	0x00007fff6fc36160 mach_msg + 60
2   com.apple.audio.CoreAudio     	0x00007fff35444bf1 HALB_MachPort::SendSimpleMessageWithSimpleReply(unsigned int, unsigned int, int, int&, bool, unsigned int) + 111
3   com.apple.audio.CoreAudio     	0x00007fff35340d88 invocation function for block in HALC_ProxyIOContext::HALC_ProxyIOContext(unsigned int, unsigned int) + 2677
4   com.apple.audio.CoreAudio     	0x00007fff35474b9e HALB_IOThread::Entry(void*) + 72
5   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
6   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 14:
0   libsystem_kernel.dylib        	0x00007fff6fc35e3e semaphore_timedwait_trap + 10
1   libdispatch.dylib             	0x00007fff6fa9bb6b _dispatch_sema4_timedwait + 76
2   libdispatch.dylib             	0x00007fff6fa9bf97 _dispatch_semaphore_wait_slow + 58
3   libdispatch.dylib             	0x00007fff6faa9795 _dispatch_worker_thread + 284
4   libsystem_pthread.dylib       	0x00007fff6fcfd109 _pthread_start + 148
5   libsystem_pthread.dylib       	0x00007fff6fcf8b8b thread_start + 15

Thread 0 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000000  rbx: 0x00007fa260a141c0  rcx: 0x000000000000000e  rdx: 0x000000000000000c
  rdi: 0x00007fa26615c400  rsi: 0x000000000000fc00  rbp: 0x00007ffee4e434c0  rsp: 0x00007ffee4e43390
   r8: 0x00007ffee4e42fa0   r9: 0x00007ffee4e43220  r10: 0x000000010afb4c5d  r11: 0x000000010afb4c68
  r12: 0x00007fa26615c400  r13: 0x00007ffee4e43398  r14: 0x0000600002c7e040  r15: 0x000000006415fcd3
  rip: 0x000000010aec56a6  rfl: 0x0000000000010202  cr2: 0x0000000000000020
  
Logical CPU:     0
Error Code:      0x00000004 (no mapping for user data read)
Trap Number:     14

**Comment by ckerr** [comment]
Converted to a bug report @ https://github.com/transmission/transmission/issues/5255
